### PR TITLE
Fix Issue 15096 - std.array.array cannot be instantiated for pointers to ranges

### PR DESCRIPTION
**Pull request by RazvanN7**
I added an overload for array which accepts pointers to iterable ranges and calls the existing overload with the value of the pointer.

### CODE REVIEW

**Comment by JackStouffer** [comment]
`refRange` exists for this exact purpose. I'd rather avoid adding pointer specializations to everything.

**Comment by MetaLang** [comment]
On the other hand, according to `isInputRange`, a pointer to a range is a range, so we should support it.

**Comment by RazvanN7** [comment]
@JackStouffer 

>  I'd rather avoid adding pointer specializations to everything.

The function should have been designed to accept pointer to ranges from the start, since pointers to ranges are ranges. I thought that overloading is the minimum invasive way to solve it, rather then messing with the existing overload

File: std/array.d [hunk 1]
```diff
@@ -157,6 +164,18 @@ if (isIterable!Range && !isNarrowString!Range && !isInfinite!Range)
...
+@safe pure nothrow unittest
+{
+    struct MyRange
+    {
+        enum front = 123;
+        enum empty = true;
+        void popFront() {}
+    }
+
+    static assert(__traits(compiles, (new MyRange).array));
```
Comment (MetaLang) [inline std/array.d]: You could stand to have a few more unit tests

**Comment by MetaLang** [comment]
@JackStouffer That's not possible in this case because a pointer to a range fails the `isIterable` check.

**Comment by JackStouffer** [comment]
Right, I'm suggesting that the range constraint be changed and static ifs used internally.

**Comment by MetaLang** [comment]
That was my first thought as well but as @RazvanN7 said, we want these changes to be unintrusive and this is the best way to do that. It can always be changed in a follow-up PR.

File: std/array.d (hunk 1 above)
Comment (wilzbach) [inline std/array.d]: I don't think that we need compiled here, just directly check the result.

**Comment by RazvanN7** [comment]
@wilzbach I added another unittest and got rid of the static assert, but I didn't do the `static if`s since in my opinion it makes the code a lot uglier and more difficult to read.

**Pull request merged**